src/parser.rs

```diff
@@ -1,7 +1,45 @@
 use std::any::Any;
 
+#[derive(Clone, Debug)]
+pub struct Input<'a> {
+    cursor: usize,
+    value: &'a str,
+}
+
+impl<'a> Input<'a> {
```

Comment:
Still missing some utility, it's currently only possible to advance until some pattern. You should also be able to read a number of chars and or similar.